aiida/engine/daemon/execmanager.py

```diff
@@ -162,21 +162,44 @@ def upload_calculation(node, transport, calc_info, script_filename, dry_run=Fals
     remote_symlink_list = calc_info.remote_symlink_list or []
 
     for uuid, filename, target in local_copy_list:
-        logger.debug('[submission of calculation {}] copying local file/folder to {}'.format(node.pk, target))
+        logger.debug('[submission of calculation {}] copying local file/folder to {}'.format(node.uuid, target))
+
+        def find_data_node(inputs, uuid):
+            """Find and return the node with the given UUID from a nested mapping of input nodes.
+
+            :param inputs: (nested) mapping of nodes
+            :param uuid: UUID of the node to find
+            :return: instance of `Node` or `None` if not found
+            """
+            from collections import Mapping
+            data_node = None
+
+            for link_label, input_node in inputs.items():
+                if isinstance(input_node, Mapping):
+                    data_node = find_data_node(input_node, uuid)
+                elif isinstance(input_node, Node) and input_node.uuid == uuid:
+                    data_node = input_node
+                if data_node is not None:
+                    break
+
+            return data_node
 
         try:
             data_node = load_node(uuid=uuid)
         except exceptions.NotExistent:
-            logger.warning('failed to load Node<{}> specified in the `local_copy_list`'.format(uuid))
+            data_node = find_data_node(inputs, uuid)
```

Comment:
do we want to allow this in general or only when `dry_run=True`?

Also, maybe good to add a comment line to both 188 and 190

Reply:
Yeah, I was wondering about that too, but I just kept the current behavior for now where it just emits a warning log. We might want to change this to actually raise. If the plugin specified that a the contents of a node need to be copied over and it is not there, that should be a problem and most likely the calculation should not work. I am not sure in what use case a mere warning is the desired alternative.